File: backend/src/main/java/com/technicalescaperoom/backend/service/admin/EventContentValidationService.java
package com.technicalescaperoom.backend.service.admin;

import com.technicalescaperoom.backend.dto.admin.EventReadinessDto;
import com.technicalescaperoom.backend.dto.admin.LevelReadinessSummaryDto;
import com.technicalescaperoom.backend.entity.*;
import com.technicalescaperoom.backend.enums.QuestionPlayer;
import com.technicalescaperoom.backend.exception.ResourceNotFoundException;
import com.technicalescaperoom.backend.repository.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class EventContentValidationService {

    private final EventRepository eventRepository;
    private final LevelRepository levelRepository;
    private final QuestionRepository questionRepository;
    private final HintRepository hintRepository;

    @Transactional(readOnly = true)
    public EventReadinessDto validateEventReadiness(Long eventId) {
        Event event = eventRepository.findById(eventId)
                .orElseThrow(() -> new ResourceNotFoundException("Event not found for ID " + eventId));

        List<String> errors = new ArrayList<>();
        Map<Integer, LevelReadinessSummaryDto> levelSummaries = new HashMap<>();

        List<Level> levels = levelRepository.findAll();
        boolean levelsCountValid = levels.stream().filter(level -> level.getLevelNumber() >= 1 && level.getLevelNumber() <= 6).count() == 6;

        if (!levelsCountValid) {
            errors.add("Event configuration requires exactly 6 levels (found: " + levels.size() + ").");
        }

        boolean allQuestionsOk = true;
        boolean allAnswersOk = true;
        boolean allHintsOk = true;

        for (int i = 1; i <= 6; i++) {
            final int levelNum = i;
            Level level = levels.stream()
                    .filter(l -> l.getLevelNumber() == levelNum)
                    .findFirst()
                    .orElse(null);

            if (level == null) {
                errors.add("Level " + levelNum + " is missing from event configuration.");
                allQuestionsOk = false;
                allAnswersOk = false;
                allHintsOk = false;
                levelSummaries.put(levelNum, LevelReadinessSummaryDto.builder()
                        .levelNumber(levelNum)
                        .levelName("Level " + levelNum)
                        .player1QuestionConfigured(false)
                        .player2QuestionConfigured(false)
                        .player1AnswerConfigured(false)
                        .player2AnswerConfigured(false)
                        .hintConfigured(false)
                        .levelReady(false)
                        .build());
                continue;
            }

                List<Question> questions = questionRepository.findByLevelIdAndIsActiveTrue(level.getId());
                Set<Integer> stages = questions.stream().map(Question::getStageNumber).collect(Collectors.toSet());
            List<Hint> hints = hintRepository.findByLevelIdOrderByDisplayOrderAsc(level.getId());

                boolean stageGraphValid = stages.size() >= 2 && stages.size() <= 3
                    && stages.stream().sorted().toList().equals(java.util.stream.IntStream.rangeClosed(1, stages.size()).boxed().toList());
                    boolean stageCoverageValid = stageGraphValid && stages.stream().allMatch(stage -> {
                    List<Question> stageQuestions = questions.stream().filter(q -> q.getStageNumber().equals(stage)).toList();
                    return stageQuestions.size() == 2
                        && stageQuestions.stream().filter(q -> q.getPlayerNumber() == QuestionPlayer.PLAYER_1).count() == 1
                        && stageQuestions.stream().filter(q -> q.getPlayerNumber() == QuestionPlayer.PLAYER_2).count() == 1;
                    });
                    boolean p1Q = stageCoverageValid && questions.stream().filter(q -> q.getPlayerNumber() == QuestionPlayer.PLAYER_1)
                        .allMatch(q -> q.getEvidence() != null && !q.getEvidence().isBlank());
                    boolean p2Q = stageCoverageValid && questions.stream().filter(q -> q.getPlayerNumber() == QuestionPlayer.PLAYER_2)
                        .allMatch(q -> q.getEvidence() != null && !q.getEvidence().isBlank());
                    boolean p1A = stageCoverageValid && questions.stream().filter(q -> q.getPlayerNumber() == QuestionPlayer.PLAYER_1)
                        .allMatch(q -> q.getExpectedAnswerHash() != null && !q.getExpectedAnswerHash().isBlank());
                    boolean p2A = stageCoverageValid && questions.stream().filter(q -> q.getPlayerNumber() == QuestionPlayer.PLAYER_2)
                        .allMatch(q -> q.getExpectedAnswerHash() != null && !q.getExpectedAnswerHash().isBlank());

            boolean hintOk = hints.stream().anyMatch(h -> Boolean.TRUE.equals(h.getIsActive()) && h.getHintContent() != null && !h.getHintContent().isBlank());

            if (!p1Q) {
                errors.add("Level " + levelNum + ": Player 1 question is missing or empty.");
                allQuestionsOk = false;
            }
            if (!p2Q) {
                errors.add("Level " + levelNum + ": Player 2 question is missing or empty.");
                allQuestionsOk = false;
            }
            if (!p1A) {
                errors.add("Level " + levelNum + ": Player 1 expected answer is missing.");
                allAnswersOk = false;
            }
            if (!p2A) {
                errors.add("Level " + levelNum + ": Player 2 expected answer is missing.");
                allAnswersOk = false;
            }
            if (!hintOk) {
                errors.add("Level " + levelNum + ": Progressive hint is missing or empty.");
                allHintsOk = false;
            }

            boolean levelReady = p1Q && p2Q && p1A && p2A && hintOk;
            levelSummaries.put(levelNum, LevelReadinessSummaryDto.builder()
                    .levelNumber(levelNum)
                    .levelName(level.getName())
                    .player1QuestionConfigured(p1Q)
                    .player2QuestionConfigured(p2Q)
                    .player1AnswerConfigured(p1A)
                    .player2AnswerConfigured(p2A)
                    .hintConfigured(hintOk)
                    .levelReady(levelReady)
                    .build());
        }

        // Final Passkey validation
        boolean passkeyConfigured = event.getPasskeyHash() != null && !event.getPasskeyHash().isBlank();
        if (!passkeyConfigured) {
            errors.add("Final Terminal configuration: Secret 6-digit passkey hash is missing.");
        }

        boolean overallReady = levelsCountValid && allQuestionsOk && allAnswersOk && allHintsOk && passkeyConfigured;

        log.info("Event #{} validation complete. Overall status: {}. Errors: {}", eventId, overallReady ? "READY" : "NOT_READY", errors.size());

        return EventReadinessDto.builder()
                .eventId(event.getId())
                .eventName(event.getName())
                .eventStatus(event.getStatus().name())
                .overallReady(overallReady)
                .levelsReady(levelsCountValid)
                .questionsReady(allQuestionsOk)
                .answersReady(allAnswersOk)
                .hintsReady(allHintsOk)
                .passkeyReady(passkeyConfigured)
                .validationErrors(errors)
                .levelSummaries(levelSummaries)
                .build();
    }
}

File: backend/services/admin/eventContentValidationService.js
const eventRepository = require('../../repositories/eventRepository');
const levelRepository = require('../../repositories/levelRepository');
const questionRepository = require('../../repositories/questionRepository');
const hintRepository = require('../../repositories/hintRepository');
const { ResourceNotFoundException } = require('../../middleware/errorHandler');

class EventContentValidationService {
  async validateEventReadiness(eventId) {
    const event = await eventRepository.findById(eventId);
    if (!event) {
      throw new ResourceNotFoundException(`Event not found for ID ${eventId}`);
    }

    const errors = [];
    const levelSummaries = {};

    const levels = await levelRepository.findAllActive();
    const validLevelNumbers = levels.filter(l => l.levelNumber >= 1 && l.levelNumber <= 6);
    const levelsCountValid = validLevelNumbers.length === 6;

    if (!levelsCountValid) {
      errors.push(`Event configuration requires exactly 6 levels (found: ${levels.length}).`);
    }

    let allQuestionsOk = true;
    let allAnswersOk = true;
    let allHintsOk = true;

    for (let i = 1; i <= 6; i++) {
      const levelNum = i;
      const level = levels.find(l => l.levelNumber === levelNum);

      if (!level) {
        errors.push(`Level ${levelNum} is missing from event configuration.`);
        allQuestionsOk = false;
        allAnswersOk = false;
        allHintsOk = false;
        levelSummaries[levelNum] = {
          levelNumber: levelNum,
          levelName: `Level ${levelNum}`,
          player1QuestionConfigured: false,
          player2QuestionConfigured: false,
          player1AnswerConfigured: false,
          player2AnswerConfigured: false,
          hintConfigured: false,
          levelReady: false
        };
        continue;
      }

      const questions = await questionRepository.findByLevelId(level.id);
      const stages = Array.from(new Set(questions.map(q => q.stageNumber))).sort((a, b) => a - b);
      const hints = await hintRepository.findByLevelIdOrderByDisplayOrderAsc(level.id);

      const stageGraphValid = stages.length >= 2 && stages.length <= 3;
      const stageCoverageValid = stageGraphValid && stages.every(stage => {
        const stageQuestions = questions.filter(q => q.stageNumber === stage);
        const p1Count = stageQuestions.filter(q => q.playerNumber === 'PLAYER_1' || q.playerNumber === 1).length;
        const p2Count = stageQuestions.filter(q => q.playerNumber === 'PLAYER_2' || q.playerNumber === 2).length;
        return stageQuestions.length === 2 && p1Count === 1 && p2Count === 1;
      });

      const p1Questions = questions.filter(q => q.playerNumber === 'PLAYER_1' || q.playerNumber === 1);
      const p2Questions = questions.filter(q => q.playerNumber === 'PLAYER_2' || q.playerNumber === 2);

      const p1Q = stageCoverageValid && p1Questions.every(q => q.evidence && q.evidence.trim().length > 0);
      const p2Q = stageCoverageValid && p2Questions.every(q => q.evidence && q.evidence.trim().length > 0);
      const p1A = stageCoverageValid && p1Questions.every(q => q.expectedAnswerHash && q.expectedAnswerHash.trim().length > 0);
      const p2A = stageCoverageValid && p2Questions.every(q => q.expectedAnswerHash && q.expectedAnswerHash.trim().length > 0);

      const hintOk = hints.some(h => h.isActive && h.hintContent && h.hintContent.trim().length > 0);

      if (!p1Q) {
        errors.push(`Level ${levelNum}: Player 1 question is missing or empty.`);
        allQuestionsOk = false;
      }
      if (!p2Q) {
        errors.push(`Level ${levelNum}: Player 2 question is missing or empty.`);
        allQuestionsOk = false;
      }
      if (!p1A) {
        errors.push(`Level ${levelNum}: Player 1 expected answer is missing.`);
        allAnswersOk = false;
      }
      if (!p2A) {
        errors.push(`Level ${levelNum}: Player 2 expected answer is missing.`);
        allAnswersOk = false;
      }
      if (!hintOk) {
        errors.push(`Level ${levelNum}: Progressive hint is missing or empty.`);
        allHintsOk = false;
      }

      const levelReady = p1Q && p2Q && p1A && p2A && hintOk;
      levelSummaries[levelNum] = {
        levelNumber: levelNum,
        levelName: level.name,
        player1QuestionConfigured: p1Q,
        player2QuestionConfigured: p2Q,
        player1AnswerConfigured: p1A,
        player2AnswerConfigured: p2A,
        hintConfigured: hintOk,
        levelReady
      };
    }

    // Final passkey validation
    const passkeyConfigured = Boolean(event.passkeyHash && event.passkeyHash.trim().length > 0);
    if (!passkeyConfigured) {
      errors.push('Final Terminal configuration: Secret 6-digit passkey hash is missing.');
    }

    const overallReady = levelsCountValid && allQuestionsOk && allAnswersOk && allHintsOk && passkeyConfigured;

    return {
      eventId: event.id,
      eventName: event.name,
      eventStatus: event.status,
      overallReady,
      levelsReady: levelsCountValid,
      questionsReady: allQuestionsOk,
      answersReady: allAnswersOk,
      hintsReady: allHintsOk,
      passkeyReady: passkeyConfigured,
      validationErrors: errors,
      levelSummaries
    };
  }
}

module.exports = new EventContentValidationService();
